### PR TITLE
fix(buildpack): use poetry versioning for python

### LINKED ISSUE #17124: Unsupported wildcard for Python version in Poetry config
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Please select which platform you are using if self-hosting.

_No response_

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

My Poetry project was configured to use Python `3.9.*` and Renovate is scheduled to run weekly on Wednesday afternoon. Last week it successfully updated dependencies in the repo's `pyproject.toml` and `poetry.lock` files without issue, but this week the pull request failed to update `poetry.lock`. From the log output in the pull request's description (see attached log), it appears that it installed Python 3.10 (specifically 3.10.6) instead of Python 3.9.

A relevant snippet from the affected `pyproject.toml` (note that the version uses an asterisk wildcard):

```toml
# ...

[tool.poetry.dependencies]
python = "3.9.*"

# ...
```

However, when I changed it to the logically-equivalent `pyproject.toml` configuration below, Renovate correctly updated `poetry.lock` (note that the version no longer uses an asterisk wildcard):

```toml
# ...

[tool.poetry.dependencies]
python = "^3.9.0"

# ...
```

This is an acceptable workaround for me, but Python version string parsing should support asterisk wildcards.

### Relevant debug logs

<details><summary>Logs</summary>

```
installing v2 tool python v3.10.6
Using prebuild python
Requirement already satisfied: pip in /opt/buildpack/tools/python/3.10.6/lib/python3.10/site-packages (22.2.1)
Collecting pip
  Downloading pip-22.2.2-py3-none-any.whl (2.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 17.3 MB/s eta 0:00:00
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 22.2.1
    Uninstalling pip-22.2.1:
      Successfully uninstalled pip-22.2.1
Successfully installed pip-22.2.2
Files removed: 2
linking tool python v3.10.6
Python 3.10.6
pip 22.2.2 from /opt/buildpack/tools/python/3.10.6/lib/python3.10/site-packages/pip (python 3.10)
Installed v2 /usr/local/buildpack/tools/v2/python.sh in 21 seconds
Collecting poetry
  Downloading poetry-1.1.14-py2.py3-none-any.whl (175 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 175.1/175.1 kB 8.6 MB/s eta 0:00:00
Collecting html5lib<2.0,>=1.0
  Downloading html5lib-1.1-py2.py3-none-any.whl (112 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 112.2/112.2 kB 14.0 MB/s eta 0:00:00
Collecting cleo<0.9.0,>=0.8.1
  Downloading cleo-0.8.1-py2.py3-none-any.whl (21 kB)
Collecting requests<3.0,>=2.18
  Downloading requests-2.28.1-py3-none-any.whl (62 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.8/62.8 kB 9.0 MB/s eta 0:00:00
Collecting virtualenv<21.0.0,>=20.0.26
  Downloading virtualenv-20.16.3-py2.py3-none-any.whl (8.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.8/8.8 MB 23.6 MB/s eta 0:00:00
Collecting cachy<0.4.0,>=0.3.0
  Downloading cachy-0.3.0-py2.py3-none-any.whl (20 kB)
Collecting pexpect<5.0.0,>=4.7.0
  Downloading pexpect-4.8.0-py2.py3-none-any.whl (59 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.0/59.0 kB 7.5 MB/s eta 0:00:00
Collecting tomlkit<1.0.0,>=0.7.0
  Downloading tomlkit-0.11.3-py3-none-any.whl (35 kB)
Collecting clikit<0.7.0,>=0.6.2
  Downloading clikit-0.6.2-py2.py3-none-any.whl (91 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 91.8/91.8 kB 12.7 MB/s eta 0:00:00
Collecting cachecontrol[filecache]<0.13.0,>=0.12.9
  Downloading CacheControl-0.12.11-py2.py3-none-any.whl (21 kB)
Collecting requests-toolbelt<0.10.0,>=0.9.1
  Downloading requests_toolbelt-0.9.1-py2.py3-none-any.whl (54 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.3/54.3 kB 7.6 MB/s eta 0:00:00
Collecting crashtest<0.4.0,>=0.3.0
  Downloading crashtest-0.3.1-py3-none-any.whl (7.0 kB)
Collecting pkginfo<2.0,>=1.4
  Downloading pkginfo-1.8.3-py2.py3-none-any.whl (26 kB)
Collecting shellingham<2.0,>=1.1
  Downloading shellingham-1.5.0-py2.py3-none-any.whl (9.3 kB)
Collecting packaging<21.0,>=20.4
  Downloading packaging-20.9-py2.py3-none-any.whl (40 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.9/40.9 kB 5.4 MB/s eta 0:00:00
Collecting keyring>=21.2.0
  Downloading keyring-23.8.2-py3-none-any.whl (34 kB)
Collecting poetry-core<1.1.0,>=1.0.7
  Downloading poetry_core-1.0.8-py2.py3-none-any.whl (425 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 425.0/425.0 kB 19.1 MB/s eta 0:00:00
Collecting msgpack>=0.5.2
  Downloading msgpack-1.0.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (316 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 317.0/317.0 kB 26.9 MB/s eta 0:00:00
Collecting lockfile>=0.9
  Downloading lockfile-0.12.2-py2.py3-none-any.whl (13 kB)
Collecting pylev<2.0,>=1.3
  Downloading pylev-1.4.0-py2.py3-none-any.whl (6.1 kB)
Collecting pastel<0.3.0,>=0.2.0
  Downloading pastel-0.2.1-py2.py3-none-any.whl (6.0 kB)
Collecting webencodings
  Downloading webencodings-0.5.1-py2.py3-none-any.whl (11 kB)
Collecting six>=1.9
  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting SecretStorage>=3.2
  Downloading SecretStorage-3.3.2-py3-none-any.whl (15 kB)
Collecting jeepney>=0.4.2
  Downloading jeepney-0.8.0-py3-none-any.whl (48 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.4/48.4 kB 5.8 MB/s eta 0:00:00
Collecting pyparsing>=2.0.2
  Downloading pyparsing-3.0.9-py3-none-any.whl (98 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.3/98.3 kB 13.2 MB/s eta 0:00:00
Collecting ptyprocess>=0.5
  Downloading ptyprocess-0.7.0-py2.py3-none-any.whl (13 kB)
Collecting idna<4,>=2.5
  Downloading idna-3.3-py3-none-any.whl (61 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.2/61.2 kB 8.0 MB/s eta 0:00:00
Collecting certifi>=2017.4.17
  Downloading certifi-2022.6.15-py3-none-any.whl (160 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 160.2/160.2 kB 19.0 MB/s eta 0:00:00
Collecting urllib3<1.27,>=1.21.1
  Downloading urllib3-1.26.11-py2.py3-none-any.whl (139 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.9/139.9 kB 17.9 MB/s eta 0:00:00
Collecting charset-normalizer<3,>=2
  Downloading charset_normalizer-2.1.0-py3-none-any.whl (39 kB)
Collecting filelock<4,>=3.4.1
  Downloading filelock-3.8.0-py3-none-any.whl (10 kB)
Collecting distlib<1,>=0.3.5
  Downloading distlib-0.3.5-py2.py3-none-any.whl (466 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 467.0/467.0 kB 17.6 MB/s eta 0:00:00
Collecting platformdirs<3,>=2.4
  Downloading platformdirs-2.5.2-py3-none-any.whl (14 kB)
Collecting cryptography>=2.0
  Downloading cryptography-37.0.4-cp36-abi3-manylinux_2_24_x86_64.whl (4.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.1/4.1 MB 27.4 MB/s eta 0:00:00
Collecting cffi>=1.12
  Downloading cffi-1.15.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (441 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 441.8/441.8 kB 20.2 MB/s eta 0:00:00
Collecting pycparser
  Downloading pycparser-2.21-py2.py3-none-any.whl (118 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 118.7/118.7 kB 14.9 MB/s eta 0:00:00
Installing collected packages: webencodings, pylev, ptyprocess, msgpack, lockfile, distlib, urllib3, tomlkit, six, shellingham, pyparsing, pycparser, poetry-core, platformdirs, pkginfo, pexpect, pastel, jeepney, idna, filelock, crashtest, charset-normalizer, certifi, cachy, virtualenv, requests, packaging, html5lib, clikit, cffi, requests-toolbelt, cryptography, cleo, cachecontrol, SecretStorage, keyring, poetry
Successfully installed SecretStorage-3.3.2 cachecontrol-0.12.11 cachy-0.3.0 certifi-2022.6.15 cffi-1.15.1 charset-normalizer-2.1.0 cleo-0.8.1 clikit-0.6.2 crashtest-0.3.1 cryptography-37.0.4 distlib-0.3.5 filelock-3.8.0 html5lib-1.1 idna-3.3 jeepney-0.8.0 keyring-23.8.2 lockfile-0.12.2 msgpack-1.0.4 packaging-20.9 pastel-0.2.1 pexpect-4.8.0 pkginfo-1.8.3 platformdirs-2.5.2 poetry-1.1.14 poetry-core-1.0.8 ptyprocess-0.7.0 pycparser-2.21 pylev-1.4.0 pyparsing-3.0.9 requests-2.28.1 requests-toolbelt-0.9.1 shellingham-1.5.0 six-1.16.0 tomlkit-0.11.3 urllib3-1.26.11 virtualenv-20.16.3 webencodings-0.5.1
The currently activated Python version 3.10.6 is not supported by the project (3.9.*).
Trying to find and use a compatible version. 

  NoCompatiblePythonVersionFound

  Poetry was unable to find a compatible version. If you have one, you can explicitly use it via the "env use" command.

  at ~/.local/lib/python3.10/site-packages/poetry/utils/env.py:768 in create_venv
       764│                     python_minor = ".".join(python_patch.split(".")[:2])
       765│                     break
       766│ 
       767│             if not executable:
    →  768│                 raise NoCompatiblePythonVersionFound(
       769│                     self._poetry.package.python_versions
       770│                 )
       771│ 
       772│         if root_venv:

```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by viceice**
## Changes
use poetry versioning instead of pep440

## Context
- closes #17124

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/exec/buildpack.spec.ts [hunk 1]
```diff
@@ -127,14 +118,38 @@ describe('util/exec/buildpack', () => {
...
     });
+
+    it.each`
+      version         | expected
+      ${'^3.9.0'}     | ${'3.10.4'}
+      ${'^3.9'}       | ${'3.10.4'}
+      ${'3.9.*'}      | ${'3.9.1'}
+      ${'3.9.1'}      | ${'3.9.1'}
+      ${'>3.8,<3.10'} | ${'3.9.1'}
+      ${'==3.9.1'}    | ${'3.10.4'}
```
Comment (viceice) [inline lib/util/exec/buildpack.spec.ts:136]: this case needs custom python versioning, which needs more time and isn't commonly used

**Review by rarkins — APPROVED** [review]

File: lib/util/exec/buildpack.ts [hunk 2]
```diff
@@ -93,7 +94,7 @@ const allToolConfig: Record<string, ToolConfig> = {
   python: {
     datasource: 'github-releases',
     depName: 'containerbase/python-prebuild',
-    versioning: pep440VersioningId,
+    versioning: poetryVersioningId,
```
Comment (steved) [inline lib/util/exec/buildpack.ts:97]: Sorry for the drive-by, was this meant to update all python versioning to poetry? This appears to break the following constraint in a Pipfile:

```
[requires]
python_version = "3.9"
```

As it's [parsed to](https://github.com/renovatebot/renovate/blob/b24d79662381764c7f51583c1feffe3632cb36b9/lib/modules/manager/pipenv/artifacts.ts#L34) `==3.9.*` which is part of PEP440.

Comment (viceice) [reply lib/util/exec/buildpack.ts:97]: will fix that today

Comment (viceice) [reply lib/util/exec/buildpack.ts:97]: - #17147
